File: sdk/AuthSys-Unity-Example/src/AuthSys.cs
using System;
using System.Collections;
using System.Text;
using UnityEngine;
using UnityEngine.Networking;

namespace AuthSys
{
    public class AuthSysClient
    {
        private string _appSecret;
        private string _version;
        private string _apiUrl;

        public string SessionToken { get; private set; }
        public string LastError { get; private set; }
        public string LastResponse { get; private set; }
        public bool Initialized { get; private set; }
        public string Username { get; private set; }
        public string Email { get; private set; }

        public AuthSysClient(string appSecret, string version, string apiUrl = "https://authsys-main-production.up.railway.app/api/v1")
        {
            _appSecret = appSecret;
            _version = version;
            _apiUrl = apiUrl.TrimEnd('/');
        }

        private string GetHWID()
        {
            try
            {
                return SystemInfo.deviceUniqueIdentifier;
            }
            catch
            {
                return "UNITY_FALLBACK_" + Application.productName;
            }
        }

        private string GetJson(string key, string json)
        {
            try
            {
                var doc = new JSONObject(json);
                if (doc.HasField(key))
                {
                    var val = doc[key];
                    if (val.IsString) return val.str;
                    return val.ToString();
                }
            }
            catch { }
            return "";
        }

        public IEnumerator InitAsync(string appName, Action<bool> callback)
        {
            LastError = "";
            LastResponse = "";
            Initialized = false;

            var json = $"{{\"app_secret\":\"{_appSecret}\",\"version\":\"{_version}\",\"hwid\":\"{GetHWID()}\",\"app_name\":\"{appName}\"}}";
            yield return PostAsync("init", json, null, (response) =>
            {
                LastResponse = response;
                var status = GetJson("status", response);
                if (status == "success" || status == "update_available")
                {
                    Initialized = true;
                    callback(true);
                }
                else
                {
                    LastError = GetJson("detail", response);
                    if (string.IsNullOrEmpty(LastError)) LastError = "Init failed";
                    callback(false);
                }
            });
        }

        public IEnumerator LoginAsync(string username, string password, Action<bool> callback, int sessionLength = 86400)
        {
            SessionToken = null;
            LastError = "";
            LastResponse = "";

            var json = $"{{\"app_secret\":\"{_appSecret}\",\"username\":\"{username}\",\"password\":\"{password}\",\"hwid\":\"{GetHWID()}\",\"session_length\":{sessionLength}}}";
            yield return PostAsync("login", json, null, (response) =>
            {
                LastResponse = response;
                var detail = GetJson("detail", response);
                if (!string.IsNullOrEmpty(detail)) { LastError = detail; callback(false); return; }

                var success = GetJson("success", response);
                if (success == "true")
                {
                    SessionToken = GetJson("token", response);
                    Username = username;
                    Email = GetJson("email", response);
                    callback(true);
                }
                else
                {
                    LastError = "Login failed";
                    callback(false);
                }
            });
        }

        public IEnumerator RegisterAsync(string username, string password, string licenseKey, Action<bool> callback, string email = "")
        {
            LastError = "";
            LastResponse = "";

            var json = $"{{\"app_secret\":\"{_appSecret}\",\"username\":\"{username}\",\"password\":\"{password}\",\"license_key\":\"{licenseKey}\",\"hwid\":\"{GetHWID()}\"";
            if (!string.IsNullOrEmpty(email)) json += $",\"email\":\"{email}\"";
            json += "}";

            yield return PostAsync("register", json, null, (response) =>
            {
                LastResponse = response;
                var detail = GetJson("detail", response);
                if (!string.IsNullOrEmpty(detail)) { LastError = detail; callback(false); return; }

                var success = GetJson("success", response);
                if (success == "true") callback(true);
                else { LastError = "Registration failed"; callback(false); }
            });
        }

        public IEnumerator LicenseLoginAsync(string licenseKey, Action<bool> callback, int sessionLength = 86400)
        {
            SessionToken = null;
            LastError = "";
            LastResponse = "";

            var json = $"{{\"app_secret\":\"{_appSecret}\",\"license_key\":\"{licenseKey}\",\"hwid\":\"{GetHWID()}\",\"session_length\":{sessionLength}}}";
            yield return PostAsync("license_login", json, null, (response) =>
            {
                LastResponse = response;
                var detail = GetJson("detail", response);
                if (!string.IsNullOrEmpty(detail)) { LastError = detail; callback(false); return; }

                var success = GetJson("success", response);
                if (success == "true")
                {
                    SessionToken = GetJson("token", response);
                    Username = GetJson("username", response);
                    callback(true);
                }
                else
                {
                    LastError = "License login failed";
                    callback(false);
                }
            });
        }

        public IEnumerator LicenseCheckAsync(string licenseKey, Action<bool> callback)
        {
            LastError = "";
            LastResponse = "";
            var json = $"{{\"app_secret\":\"{_appSecret}\",\"license_key\":\"{licenseKey}\"}}";
            yield return PostAsync("license/check", json, null, (response) =>
            {
                LastResponse = response;
                callback(true);
            });
        }

        public IEnumerator VerifyAsync(Action<bool> callback)
        {
            LastError = "";
            LastResponse = "";
            if (string.IsNullOrEmpty(SessionToken)) { LastError = "No active session"; callback(false); yield break; }
            yield return PostAsync("verify", "{}", SessionToken, (response) =>
            {
                LastResponse = response;
                callback(true);
            });
        }

        public IEnumerator ChatSendAsync(int roomId, string message, Action<bool> callback)
        {
            LastError = "";
            LastResponse = "";
            var endpoint = $"chat/send?room_id={roomId}&message={UnityEngine.Networking.UnityWebRequest.EscapeURL(message)}";
            yield return PostAsync(endpoint, "{}", SessionToken, (response) =>
            {
                LastResponse = response;
                callback(true);
            });
        }

        private IEnumerator PostAsync(string endpoint, string jsonBody, string token, Action<string> callback)
        {
            var url = $"{_apiUrl}/client/{endpoint}";
            using (var req = new UnityWebRequest(url, "POST"))
            {
                byte[] bodyRaw = Encoding.UTF8.GetBytes(jsonBody);
                req.uploadHandler = new UploadHandlerRaw(bodyRaw);
                req.downloadHandler = new DownloadHandlerBuffer();
                req.SetRequestHeader("Content-Type", "application/json");

                if (!string.IsNullOrEmpty(token))
                {
                    req.SetRequestHeader("Authorization", "Bearer " + token);
                    req.SetRequestHeader("X-HWID", GetHWID());
                }

                req.timeout = 30;
                yield return req.SendWebRequest();

                if (req.result == UnityWebRequest.Result.Success)
                {
                    callback(req.downloadHandler.text);
                }
                else
                {
                    callback($"{{\"success\":false,\"detail\":\"{req.error}\"}}");
                }
            }
        }

        public string Var(string name)
        {
            return GetJson(name, LastResponse);
        }

        public void Logout()
        {
            SessionToken = null;
            Username = null;
            Email = null;
            LastError = "";
            LastResponse = "";
        }
    }
}

File: sdk/AuthSys-Rust-Example/src/authsys.rs
use reqwest::blocking::Client;
use serde_json::{json, Value};
use std::time::Duration;

pub struct AuthSysClient {
    app_secret: String,
    version: String,
    api_url: String,
    client: Client,

    pub session_token: Option<String>,
    pub last_error: String,
    pub last_response: String,
    pub initialized: bool,
    pub username: String,
    pub email: String,
}

impl AuthSysClient {
    pub fn new(app_secret: &str, version: &str, api_url: Option<&str>) -> Self {
        Self {
            app_secret: app_secret.to_string(),
            version: version.to_string(),
            api_url: api_url.unwrap_or("https://authsys-main-production.up.railway.app/api/v1").trim_end_matches('/').to_string(),
            client: Client::builder().timeout(Duration::from_secs(30)).build().unwrap(),
            session_token: None,
            last_error: String::new(),
            last_response: String::new(),
            initialized: false,
            username: String::new(),
            email: String::new(),
        }
    }

    fn get_hwid() -> String {
        // Machine ID
        if let Ok(id) = std::fs::read_to_string("/etc/machine-id") {
            return id.trim().to_string();
        }
        // Windows volume serial
        if cfg!(target_os = "windows") {
            let output = std::process::Command::new("wmic")
                .args(&["volume", "where", "DriveLetter='C:'", "get", "SerialNumber", "/value"])
                .output();
            if let Ok(out) = output {
                let stdout = String::from_utf8_lossy(&out.stdout);
                if let Some(line) = stdout.lines().find(|l| l.contains("SerialNumber")) {
                    if let Some(val) = line.split('=').nth(1) {
                        return val.trim().to_string();
                    }
                }
            }
        }
        format!("{:x}", std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_nanos())
    }

    fn get_json(&self, key: &str) -> String {
        if let Ok(v) = serde_json::from_str::<Value>(&self.last_response) {
            if let Some(val) = v.get(key) {
                if val.is_string() { return val.as_str().unwrap_or("").to_string(); }
                return val.to_string();
            }
        }
        String::new()
    }

    fn post(&self, endpoint: &str, body: Value, token: Option<&str>) -> String {
        let url = format!("{}/client/{}", self.api_url, endpoint);
        let mut req = self.client.post(&url).json(&body).header("Content-Type", "application/json");

        if let Some(t) = token {
            req = req.header("Authorization", format!("Bearer {}", t));
            req = req.header("X-HWID", Self::get_hwid());
        }

        match req.send() {
            Ok(res) => res.text().unwrap_or_else(|e| format!("{{\"success\":false,\"detail\":\"{}\"}}", e)),
            Err(e) => format!("{{\"success\":false,\"detail\":\"{}\"}}", e),
        }
    }

    pub fn init(&mut self, app_name: Option<&str>) {
        self.last_error.clear();
        self.last_response.clear();
        self.initialized = false;

        let body = json!({
            "app_secret": self.app_secret,
            "version": self.version,
            "hwid": Self::get_hwid(),
            "app_name": app_name.unwrap_or(""),
        });
        self.last_response = self.post("init", body, None);

        let status = self.get_json("status");
        if status == "success" || status == "update_available" {
            self.initialized = true;
        } else {
            self.last_error = self.get_json("detail");
            if self.last_error.is_empty() { self.last_error = "Init failed".to_string(); }
        }
    }

    pub fn login(&mut self, username: &str, password: &str, session_length: Option<i32>) {
        self.session_token = None;
        self.last_error.clear();
        self.last_response.clear();

        let body = json!({
            "app_secret": self.app_secret,
            "username": username,
            "password": password,
            "hwid": Self::get_hwid(),
            "session_length": session_length.unwrap_or(86400),
        });
        self.last_response = self.post("login", body, None);

        let detail = self.get_json("detail");
        if !detail.is_empty() { self.last_error = detail; return; }

        let success = self.get_json("success");
        if success == "true" {
            self.session_token = Some(self.get_json("token"));
            self.username = username.to_string();
            self.email = self.get_json("email");
        } else {
            self.last_error = "Login failed".to_string();
        }
    }

    pub fn register(&mut self, username: &str, password: &str, license_key: &str, email: Option<&str>) {
        self.last_error.clear();
        self.last_response.clear();

        let mut body = json!({
            "app_secret": self.app_secret,
            "username": username,
            "password": password,
            "license_key": license_key,
            "hwid": Self::get_hwid(),
        });
        if let Some(e) = email {
            if !e.is_empty() {
                body["email"] = json!(e);
            }
        }
        self.last_response = self.post("register", body, None);

        let detail = self.get_json("detail");
        if !detail.is_empty() { self.last_error = detail; return; }

        let success = self.get_json("success");
        if success != "true" { self.last_error = "Registration failed".to_string(); }
    }

    pub fn license_login(&mut self, license_key: &str, session_length: Option<i32>) {
        self.session_token = None;
        self.last_error.clear();
        self.last_response.clear();

        let body = json!({
            "app_secret": self.app_secret,
            "license_key": license_key,
            "hwid": Self::get_hwid(),
            "session_length": session_length.unwrap_or(86400),
        });
        self.last_response = self.post("license_login", body, None);

        let detail = self.get_json("detail");
        if !detail.is_empty() { self.last_error = detail; return; }

        let success = self.get_json("success");
        if success == "true" {
            self.session_token = Some(self.get_json("token"));
            self.username = self.get_json("username");
        } else {
            self.last_error = "License login failed".to_string();
        }
    }

    pub fn license_check(&mut self, license_key: &str) {
        self.last_error.clear();
        self.last_response.clear();
        let body = json!({
            "app_secret": self.app_secret,
            "license_key": license_key,
        });
        self.last_response = self.post("license/check", body, None);
    }

    pub fn verify(&mut self) {
        self.last_error.clear();
        self.last_response.clear();
        if self.session_token.is_none() { self.last_error = "No active session".to_string(); return; }
        let body = json!({});
        self.last_response = self.post("verify", body, self.session_token.as_deref());
    }

    pub fn chat_send(&mut self, room_id: i32, message: &str) {
        self.last_error.clear();
        self.last_response.clear();
        let endpoint = format!("chat/send?room_id={}&message={}", room_id, urlencoding(message));
        let body = json!({});
        self.last_response = self.post(&endpoint, body, self.session_token.as_deref());
    }

    pub fn var(&self, name: &str) -> String {
        self.get_json(name)
    }

    pub fn logout(&mut self) {
        self.session_token = None;
        self.username.clear();
        self.email.clear();
        self.last_error.clear();
        self.last_response.clear();
    }
}

fn urlencoding(s: &str) -> String {
    s.chars().map(|c| match c {
        'A'..='Z' | 'a'..='z' | '0'..='9' | '-' | '_' | '.' | '~' => c.to_string(),
        _ => format!("%{:02X}", c as u8),
    }).collect()
}
